test/test_image.py

```diff
@@ -54,176 +53,202 @@ def normalize_dimensions(img_pil):
+])
+@pytest.mark.parametrize('conversion', [
+    (None, ImageReadMode.UNCHANGED),
+    ("L", ImageReadMode.GRAY),
+    ("RGB", ImageReadMode.RGB),
+])
+def test_decode_jpeg(img_path, conversion):
+    pil_mode, mode = conversion
+
+    with Image.open(img_path) as img:
+        is_cmyk = img.mode == "CMYK"
+        if pil_mode is not None:
+            if is_cmyk:
+                # libjpeg does not support the conversion
+                pytest.xfail("Decoding a CMYK jpeg isn't supported")
+            img = img.convert(pil_mode)
+        img_pil = torch.from_numpy(np.array(img))
+        if is_cmyk:
+            # flip the colors to match libjpeg
+            img_pil = 255 - img_pil
+
+    img_pil = normalize_dimensions(img_pil)
+    data = read_file(img_path)
+    img_ljpeg = decode_image(data, mode=mode)
+
+    # Permit a small variation on pixel values to account for implementation
+    # differences between Pillow and LibJPEG.
+    abs_mean_diff = (img_ljpeg.type(torch.float32) - img_pil).abs().mean().item()
+    assert abs_mean_diff < 2
+
+    with pytest.raises(RuntimeError, match="Expected a non empty 1-dimensional tensor"):
+        decode_jpeg(torch.empty((100, 1), dtype=torch.uint8))
+
+    with pytest.raises(RuntimeError, match="Expected a torch.uint8 tensor"):
+        decode_jpeg(torch.empty((100,), dtype=torch.float16))
+
+    with pytest.raises(RuntimeError):
+        decode_jpeg(torch.empty((100), dtype=torch.uint8))
+
+
+@pytest.mark.parametrize('img_path', [
+    pytest.param(truncated_image, id=_get_safe_image_name(truncated_image))
+    for truncated_image in glob.glob(os.path.join(DAMAGED_JPEG, 'corrupt*.jpg'))
+])
+def test_damaged_images(img_path):
+    # Test image with bad Huffman encoding (should not raise)
+    bad_huff = read_file(os.path.join(DAMAGED_JPEG, 'bad_huffman.jpg'))
+    try:
+        _ = decode_jpeg(bad_huff)
+    except RuntimeError:
+        assert False
+
+    # Truncated images should raise an exception
+    data = read_file(img_path)
+    with pytest.raises(RuntimeError):
+        decode_jpeg(data)
+
+
+@pytest.mark.parametrize('img_path', [
+    pytest.param(png_path, id=_get_safe_image_name(png_path))
+    for png_path in get_images(FAKEDATA_DIR, ".png")
+])
+@pytest.mark.parametrize('conversion', [
+    (None, ImageReadMode.UNCHANGED),
+    ("L", ImageReadMode.GRAY),
+    ("LA", ImageReadMode.GRAY_ALPHA),
+    ("RGB", ImageReadMode.RGB),
+    ("RGBA", ImageReadMode.RGB_ALPHA),
+])
+def test_decode_png(img_path, conversion):
+    pil_mode, mode = conversion
+
+    with Image.open(img_path) as img:
+        if pil_mode is not None:
+            img = img.convert(pil_mode)
+        img_pil = torch.from_numpy(np.array(img))
+
+    img_pil = normalize_dimensions(img_pil)
+    data = read_file(img_path)
+    img_lpng = decode_image(data, mode=mode)
+
+    tol = 0 if conversion is None else 1
+    assert img_lpng.allclose(img_pil, atol=tol)
+
+    with pytest.raises(RuntimeError):
```

Comment:
Nit: same here it'd be good to have the `match=`

Also let's move the assertions out in another function

Reply:
Done!